editor/find_in_files_scanner.cpp

```diff
@@ -0,0 +1,352 @@
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "find_in_files_scanner.h"
+
+const int SCRIPT_SOURCE_TAG_LEN = String("script/source = \"").length();
+
+Vector<FindInFilesScanner::ScanMatch> FindInFilesScanner::scan(String file_path, String pattern, bool match_case, bool whole_words) {
+	Ref<FileAccess> file = FileAccess::open(file_path, FileAccess::READ);
+	Vector<FindInFilesScanner::ScanMatch> matches;
+	if (file.is_null()) {
+		print_verbose(String("Cannot open file ") + file_path);
+		return matches;
+	}
+
+	String extension = file->get_path().get_extension().to_lower();
+	if (extension == "tscn") {
+		BuiltinScriptFindInFilesScanner::scan(file, file_path, pattern, match_case, whole_words, matches);
+	} else {
+		// No specific match was found, use default scanner
+		DefaultFindInFilesScanner::scan(file, file_path, file_path, pattern, match_case, whole_words, matches);
+	}
+
+	return matches;
+}
+
+void FindInFilesScanner::replace(String relative_path, String absolute_path, Vector<ScanLocation> &locations, bool match_case, bool whole_words, String search_text, String new_text) {
+	Ref<FileAccess> file = FileAccess::open(absolute_path, FileAccess::READ);
+	if (file.is_null()) {
+		print_verbose(String("Cannot open file ") + file->get_path());
+		return;
+	}
+
+	String extension = absolute_path.get_extension().to_lower();
+	if (extension == "tscn") {
+		BuiltinScriptFindInFilesScanner::replace(file, relative_path, locations, match_case, whole_words, search_text, new_text);
+	} else {
+		// No specific match was found, use default scanner
+		DefaultFindInFilesScanner::replace(file, relative_path, locations, match_case, whole_words, search_text, new_text);
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+void DefaultFindInFilesScanner::scan(Ref<FileAccess> file, String file_path, String display_text, String pattern, bool match_case, bool whole_words, Vector<FindInFilesScanner::ScanMatch> &matches, Size2i range) {
+	// Line number starts at 1.
+	for (int line_number = 1, begin = 0, end = 0; !file->eof_reached(); line_number++, begin = 0, end = 0) {
```

Comment:
`begin` and `end` should be simply declared as scope variables inside of `for`. This is unnecessarily convoluted.